### PR TITLE
REF: use public pandas API in dataframe.empty

### PR DESCRIPTION
**Pull request by jbrockmendel**
The current usage of non-public API is breaking the docbuild on a pandas PR: https://github.com/pandas-dev/pandas/pull/40149/checks?check_run_id=2102847175#step:5:152

See also: https://github.com/pandas-dev/pandas/issues/40226

A slightly more performant implementation may become possible following https://github.com/pandas-dev/pandas/issues/39776

### CODE REVIEW

**Comment by martindurant** [comment]
Thanks for taking an interest and helping out here!

Unfortunately, the existence of this function and the convoluted code it contains, is because of pandas' poor performance in forming dataframes to be assigned into. Benchmarking, I find:
```python
# main branch
In [2]: %timeit out = fastparquet.dataframe.empty('i4,u2,f4,f2,f4,M8,category', 1000000, cats={6: 5})
903 µs ± 86.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

# this branch
In [3]: %timeit out = fastparquet.dataframe.empty('i4,u2,f4,f2,f4,M8,category', 1000000, cats={6: 5})
43.6 ms ± 668 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Fastparquet does not want this factor of 50 slowdown! I really do hope that Pandas itself creates this functionality, so that we don't have to.

**Comment by jbrockmendel** [comment]
Yep, thats a pretty big slowdown.  I'll see what I can do upstream, and take another try at this in the interim.

**Comment by martindurant** [comment]
Thanks, @jbrockmendel !

Interestingly, I [wrote about this](http://martindurant.github.io/blog/pandas-and-memory-allocation/) all the way back in 2017. @jreback helped write this initially, and @TomAugspurger has helped keep this module in sync with pandas changes.

**Comment by jbrockmendel** [comment]
Updated with an implementation that slightly outperforms the status quo

```
In [3]: %timeit out = fastparquet.dataframe.empty('i4,u2,f4,f2,f4,M8,category', 1000000, cats={6: 5})
741 µs ± 11.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- main
712 µs ± 4.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR
```

This still accesses internals in a way that is not officially supported, but should be at least future-proof to https://github.com/pandas-dev/pandas/pull/40149 (though not ArrayManager)

**Comment by martindurant** [comment]
Perfect - that persuades me. I'll leave this a day or so to see if there are other ocmments.

**Comment by martindurant** [comment]
Thanks @jbrockmendel !

**Comment by jbrockmendel** [comment]
Some of what I'm working on in pandas would be simplified if this made it into a released version of fastparquet.  Is that on the horizon?

**Comment by martindurant** [comment]
Not in the next day or two, but yes.

File: fastparquet/dataframe.py [hunk 1]
```diff
@@ -163,31 +163,33 @@ def set_cats(values, i=i, col=col, **kwargs):
...
+        elif getattr(bvalues.dtype, 'tz', None):
+            values = np.empty(shape=shape, dtype='M8[ns]')
+            values = type(bvalues)(values, dtype=bvalues.dtype)
         else:
-            new_shape = (block.values.shape[0], size)
-            values = np.empty(shape=new_shape, dtype=block.values.dtype)
-            new_block = block.make_block_same_class(values=values)
+            # Note: this will break on any ExtensionDtype other than
+            #  Categorical and DatetimeTZ
+            values = np.empty(shape=shape, dtype=bvalues.dtype)
```
[the remark this answers is not in the document]
Comment (martindurant) [reply fastparquet/dataframe.py:186]: Thanks for looking into it. Happy to push a point release to fix pandas' CI, if needed.

I hope the line above doesn't make a copy!

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply fastparquet/dataframe.py:186]: if its a 3rd-party EA then there's no telling, but for the dt64 case this should be copy-free

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply fastparquet/dataframe.py:186]: reproduced the failure locally and the edit above fixes 2 of the 8 tests.  the others look pytz-related.

[the remark this answers is not in the document]
Comment (martindurant) [reply fastparquet/dataframe.py:186]: It it's a copy, fastparquet will be filling out an array that's no longer connected to the actual dataframe storage?

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply fastparquet/dataframe.py:186]: > It it's a copy, fastparquet will be filling out an array that's no longer connected to the actual dataframe storage?

No, correctness wouldn't be affected.

**Pull request merged**